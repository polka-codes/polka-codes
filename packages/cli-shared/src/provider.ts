import { spawn } from 'node:child_process'
import { mkdir, readFile, rename, unlink, writeFile } from 'node:fs/promises'
import { dirname, normalize, resolve } from 'node:path'
import { input, select } from '@inquirer/prompts'
import type { ToolProvider } from '@polka-codes/core'
import ignore from 'ignore'
import { lookup } from 'mime-types'
import { checkRipgrep } from './utils/checkRipgrep'
import { listFiles } from './utils/listFiles'
import { searchFiles } from './utils/searchFiles'

export type ProviderOptions = {
  command?: {
    onStarted(command: string): void
    onStdout(data: string): void
    onStderr(data: string): void
    onExit(code: number): void
    onError(error: unknown): void
  }
  excludeFiles?: string[]
  summarizeOutput?: (stdout: string, stderr: string) => Promise<string | undefined>
  summaryThreshold?: number
}

export const getProvider = (options: ProviderOptions = {}): ToolProvider => {
  const ig = ignore().add(options.excludeFiles ?? [])
  const memoryStore: Record<string, string> = {}
  const provider: ToolProvider = {
    listTopics: async (): Promise<string[]> => {
      return Object.keys(memoryStore)
    },
    read: async (topic: string): Promise<string | undefined> => {
      return memoryStore[topic]
    },
    append: async (topic: string, content: string): Promise<void> => {
      if (memoryStore[topic]) {
        memoryStore[topic] += content
      } else {
        memoryStore[topic] = content
      }
    },
    replace: async (topic: string, content: string): Promise<void> => {
      memoryStore[topic] = content
    },
    remove: async (topic: string): Promise<void> => {
      delete memoryStore[topic]
    },
    readFile: async (path: string, includeIgnored: boolean): Promise<string | undefined> => {
      if (!includeIgnored && ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      try {
        return await readFile(path, 'utf8')
      } catch (_e) {
        return undefined
      }
    },
    writeFile: async (path: string, content: string): Promise<void> => {
      if (ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      // generate parent directories if they don't exist
      await mkdir(dirname(path), { recursive: true })

      return await writeFile(path, content, 'utf8')
    },
    removeFile: async (path: string): Promise<void> => {
      if (ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      return await unlink(path)
    },
    renameFile: async (sourcePath: string, targetPath: string): Promise<void> => {
      if (ig.ignores(sourcePath) || ig.ignores(targetPath)) {
        throw new Error(`Not allow to access file ${sourcePath} or ${targetPath}`)
      }
      return await rename(sourcePath, targetPath)
    },
    listFiles: async (path: string, recursive: boolean, maxCount: number, includeIgnored: boolean): Promise<[string[], boolean]> => {
      return await listFiles(path, recursive, maxCount, process.cwd(), options.excludeFiles, includeIgnored)
    },
    readBinaryFile: async (url: string) => {
      if (url.startsWith('file://')) {
        const filePath = decodeURIComponent(url.substring('file://'.length))
        const resolvedPath = normalize(resolve(process.cwd(), filePath))

        if (!resolvedPath.startsWith(process.cwd())) {
          throw new Error(`Access to file path "${filePath}" is restricted.`)
        }

        const data = await readFile(resolvedPath)
        const mediaType = lookup(resolvedPath) || 'application/octet-stream'

        return {
          base64Data: data.toString('base64'),
          mediaType,
        }
      }
      const response = await fetch(url)
      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`)
      }
      const data = await response.arrayBuffer()
      const mediaType = lookup(url) || 'application/octet-stream'

      return {
        base64Data: Buffer.from(data).toString('base64'),
        mediaType,
      }
    },

    executeCommand: (
      command: string,
      _needApprove: boolean,
    ): Promise<{ stdout: string; stderr: string; exitCode: number; summary?: string }> => {
      // TODO: add timeout

      return new Promise((resolve, reject) => {
        // spawn a shell to execute the command

        options.command?.onStarted(command)

        const child = spawn(command, [], {
          shell: true,
          stdio: 'pipe',
        })

        let stdoutText = ''
        let stderrText = ''

        child.stdout.on('data', (data) => {
          const dataStr = data.toString()
          options.command?.onStdout(dataStr)
          stdoutText += dataStr
        })

        child.stderr.on('data', (data) => {
          const dataStr = data.toString()
          options.command?.onStderr(dataStr)
          stderrText += dataStr
        })

        child.on('close', async (code) => {
          options.command?.onExit(code ?? 0)
          const totalLength = stdoutText.length + stderrText.length
          if (totalLength > (options.summaryThreshold ?? 5000) && options.summarizeOutput) {
            try {
              const summary = await options.summarizeOutput(stdoutText, stderrText)
              if (summary) {
                resolve({
                  summary,
                  stdout: stdoutText,
                  stderr: stderrText,
                  exitCode: code ?? 0,
                })
                return
              }
            } catch (_e) {
              console.error('Summarization failed:', _e)
            }
          }
          resolve({
            stdout: stdoutText,
            stderr: stderrText,
            exitCode: code ?? 0,
          })
        })

        child.on('error', (err) => {
          options.command?.onError(err)
          reject(err)
        })
      })
    },
    askFollowupQuestion: async (question: string, answerOptions: string[]): Promise<string> => {
      if (answerOptions.length === 0) {
        return await input({ message: question })
      }

      const otherMessage = 'Other (enter text)'
      answerOptions.push(otherMessage)
      const answer = await select({
        message: question,
        choices: answerOptions.map((option) => ({ name: option, value: option })),
      })

      if (answer === otherMessage) {
        return await input({ message: 'Enter your answer:' })
      }
      return answer
    },
    fetchUrl: async (url: string): Promise<string> => {
      const isRaw = url.startsWith('https://raw.githubusercontent.com/')

      const urlToFetch = isRaw ? url : `https://r.jina.ai/${url}`

      try {
        const response = await fetch(urlToFetch)
        if (!response.ok) {
          throw new Error(`HTTP error! status: ${response.status}`)
        }
        return await response.text()
      } catch (error) {
        console.error('Error fetching URL:', error)
        throw error
      }
    },
  }

  if (checkRipgrep()) {
    provider.searchFiles = async (path: string, regex: string, filePattern: string): Promise<string[]> => {
      return await searchFiles(path, regex, filePattern, process.cwd(), options.excludeFiles)
    }
  } else {
    console.error(
      'Error: ripgrep (rg) is not installed. Search file tool is disabled. Please install it from https://github.com/BurntSushi/ripgrep#installation',
    )
  }

  return provider
}
