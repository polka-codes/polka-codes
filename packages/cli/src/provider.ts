import { spawn } from 'node:child_process'
import { mkdir, readFile, writeFile } from 'node:fs/promises'
import { dirname } from 'node:path'
import ignore from 'ignore'

import type { ToolProvider } from '@polka-codes/core'
import { listFiles } from './utils/listFiles'
import { searchFiles } from './utils/searchFiles'

export type ProviderOptions = {
  command: {
    onStarted(command: string): void
    onStdout(data: string): void
    onStderr(data: string): void
    onExit(code: number): void
    onError(error: unknown): void
  }
  excludeFiles?: string[]
}

export const getProvider = (options: ProviderOptions): ToolProvider => {
  const ig = ignore().add(options.excludeFiles ?? [])
  return {
    readFile: async (path: string): Promise<string> => {
      if (ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      return await readFile(path, 'utf8')
    },
    writeFile: async (path: string, content: string): Promise<void> => {
      if (ig.ignores(path)) {
        throw new Error(`Not allow to access file ${path}`)
      }
      // generate parent directories if they don't exist
      await mkdir(dirname(path), { recursive: true })

      return await writeFile(path, content, 'utf8')
    },
    listFiles: async (path: string, recursive: boolean, maxCount: number): Promise<[string[], boolean]> => {
      return await listFiles(path, recursive, maxCount, dirname(path), options.excludeFiles)
    },
    searchFiles: async (path: string, regex: string, filePattern: string): Promise<string[]> => {
      return await searchFiles(path, regex, filePattern, dirname(path), options.excludeFiles)
    },
    // listCodeDefinitionNames: async (path: string) => Promise<string[]> {},

    executeCommand: (command: string, needApprove: boolean): Promise<{ stdout: string; stderr: string; exitCode: number }> => {
      // TODO: add timeout

      return new Promise((resolve, reject) => {
        // spawn a shell to execute the command

        options.command.onStarted(command)

        const child = spawn(command, [], {
          shell: true,
          stdio: 'pipe',
        })

        let stdoutText = ''
        let stderrText = ''

        child.stdout.on('data', (data) => {
          const dataStr = data.toString()
          options.command.onStdout(dataStr)
          stdoutText += dataStr
        })

        child.stderr.on('data', (data) => {
          const dataStr = data.toString()
          options.command.onStderr(dataStr)
          stderrText += dataStr
        })

        child.on('close', (code) => {
          options.command.onExit(code ?? 0)
          resolve({
            stdout: stdoutText,
            stderr: stderrText,
            exitCode: code ?? 0,
          })
        })

        child.on('error', (err) => {
          options.command.onError(err)
          reject(err)
        })
      })
    },
    // askFollowupQuestion: async (question: string, options: string[]) => Promise<string> {},
    // attemptCompletion: async (result: string) => Promise<string | undefined> {},
  }
}
