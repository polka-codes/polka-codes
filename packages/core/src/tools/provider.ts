export type FilesystemProvider = {
  readFile?: (path: string) => Promise<string>
  writeFile?: (path: string, content: string) => Promise<void>
  removeFile?: (path: string) => Promise<void>
  listFiles?: (path: string, recursive: boolean, maxCount: number) => Promise<[string[], boolean]>
  searchFiles?: (path: string, regex: string, filePattern: string) => Promise<string[]>
  listCodeDefinitionNames?: (path: string) => Promise<string[]>
}

export type CommandProvider = {
  executeCommand?: (command: string, needApprove: boolean) => Promise<{ stdout: string; stderr: string; exitCode: number }>
}

export type InteractionProvider = {
  askFollowupQuestion?: (question: string, options: string[]) => Promise<string>
  attemptCompletion?: (result: string) => Promise<string | undefined>
}

export type ToolProvider = FilesystemProvider & CommandProvider & InteractionProvider

export class MockProvider implements ToolProvider {
  async readFile(path: string): Promise<string> {
    return 'mock content'
  }

  async writeFile(path: string, content: string): Promise<void> {
    return
  }

  async removeFile(path: string): Promise<void> {
    return
  }

  async listFiles(path: string, recursive: boolean, maxCount: number): Promise<[string[], boolean]> {
    return [['mock-file.txt'], false]
  }

  async searchFiles(path: string, regex: string, filePattern: string): Promise<string[]> {
    return ['mock-file.txt']
  }

  async listCodeDefinitionNames(path: string): Promise<string[]> {
    return ['mockDefinition']
  }

  async executeCommand(command: string, needApprove: boolean): Promise<{ stdout: string; stderr: string; exitCode: number }> {
    return { stdout: 'mock output', stderr: '', exitCode: 0 }
  }

  async askFollowupQuestion(question: string, options?: string[]): Promise<string> {
    return 'mock answer'
  }

  async attemptCompletion(result: string): Promise<string | undefined> {
    return 'mock completion'
  }
}
