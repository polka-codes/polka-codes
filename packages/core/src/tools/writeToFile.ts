import { type FullToolInfo, type ToolHandler, type ToolInfo, ToolResponseType } from '../tool'
import type { FilesystemProvider } from './provider'
import { getString } from './utils'

export const toolInfo = {
  name: 'write_to_file',
  description:
    "Request to write content to a file at the specified path. If the file exists, it will be overwritten with the provided content. If the file doesn't exist, it will be created. This tool will automatically create any directories needed to write the file.",
  parameters: [
    {
      name: 'path',
      description: 'The path of the file to write to',
      required: true,
      usageValue: 'File path here',
    },
    {
      name: 'content',
      description:
        "The content to write to the file. ALWAYS provide the COMPLETE intended content of the file, without any truncation or omissions. You MUST include ALL parts of the file, even if they haven't been modified.",
      required: true,
      usageValue: 'Your file content here',
    },
  ],
  examples: [
    {
      description: 'Request to write content to a file',
      parameters: [
        {
          name: 'path',
          value: 'src/main.js',
        },
        {
          name: 'content',
          value: `import React from 'react';

function App() {
  return (
    <div>
      <h1>Hello, World!</h1>
    </div>
  );
}

export default App;
`,
        },
      ],
    },
  ],
} as const satisfies ToolInfo

export const handler: ToolHandler<typeof toolInfo, FilesystemProvider> = async (provider, args) => {
  if (!provider.writeFile) {
    return {
      type: ToolResponseType.Error,
      message: 'Not possible to write file. Abort.',
    }
  }

  const path = getString(args, 'path')
  const content = getString(args, 'content')

  // Create parent directories if they don't exist
  await provider.writeFile(path, content)

  return {
    type: ToolResponseType.Reply,
    message: `<write_file_path>${path}</write_file_path><status>Success</status>`,
  }
}

export const isAvailable = (provider: FilesystemProvider): boolean => {
  return !!provider.writeFile
}

export default {
  ...toolInfo,
  handler,
  isAvailable,
} satisfies FullToolInfo
