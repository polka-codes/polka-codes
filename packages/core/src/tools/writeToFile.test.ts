import { describe, expect, it, spyOn } from 'bun:test'
import { MockProvider } from './provider'
import writeToFile from './writeToFile'

describe('writeToFile', () => {
  it('should write file successfully', async () => {
    const mockProvider = new MockProvider()
    spyOn(mockProvider, 'writeFile').mockResolvedValue()

    const result = await writeToFile.handler(mockProvider, {
      path: 'test.txt',
      content: 'Test content',
    })

    expect(result).toMatchSnapshot()
    expect(mockProvider.writeFile).toHaveBeenCalledWith('test.txt', 'Test content')
  })

  it('should handle directory creation', async () => {
    const mockProvider = new MockProvider()
    spyOn(mockProvider, 'writeFile').mockResolvedValue()

    const result = await writeToFile.handler(mockProvider, {
      path: 'new-dir/test.txt',
      content: 'Test content',
    })

    expect(result).toMatchSnapshot()
    expect(mockProvider.writeFile).toHaveBeenCalledWith('new-dir/test.txt', 'Test content')
  })

  it('should handle write errors', async () => {
    const mockProvider = new MockProvider()
    spyOn(mockProvider, 'writeFile').mockRejectedValue(new Error('Write error'))

    const result = writeToFile.handler(mockProvider, {
      path: 'error.txt',
      content: 'Test content',
    })

    expect(result).rejects.toMatchSnapshot()
    expect(mockProvider.writeFile).toHaveBeenCalledWith('error.txt', 'Test content')
  })

  it('should remove CDATA tags when present', async () => {
    const mockProvider = new MockProvider()
    spyOn(mockProvider, 'writeFile').mockResolvedValue()

    const result = await writeToFile.handler(mockProvider, {
      path: 'cdata.txt',
      content: '<![CDATA[Test content]]>',
    })

    expect(result).toMatchSnapshot()
    expect(mockProvider.writeFile).toHaveBeenCalledWith('cdata.txt', 'Test content')
  })

  it('should not modify content without CDATA tags', async () => {
    const mockProvider = new MockProvider()
    spyOn(mockProvider, 'writeFile').mockResolvedValue()

    const result = await writeToFile.handler(mockProvider, {
      path: 'nocdata.txt',
      content: 'Test content',
    })

    expect(result).toMatchSnapshot()
    expect(mockProvider.writeFile).toHaveBeenCalledWith('nocdata.txt', 'Test content')
  })

  it('should not modify partial CDATA tags', async () => {
    const mockProvider = new MockProvider()
    spyOn(mockProvider, 'writeFile').mockResolvedValue()

    const result = await writeToFile.handler(mockProvider, {
      path: 'partial.txt',
      content: '<![CDATA[Test content',
    })

    expect(result).toMatchSnapshot()
    expect(mockProvider.writeFile).toHaveBeenCalledWith('partial.txt', '<![CDATA[Test content')
  })
})
