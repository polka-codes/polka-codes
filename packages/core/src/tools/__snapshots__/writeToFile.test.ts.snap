// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`writeToFile should write file successfully 1`] = `
{
  "message": "<write_to_file_path>test.txt</write_to_file_path><status>Success</status>",
  "type": "Reply",
}
`;

exports[`writeToFile should handle directory creation 1`] = `
{
  "message": "<write_to_file_path>new-dir/test.txt</write_to_file_path><status>Success</status>",
  "type": "Reply",
}
`;

exports[`writeToFile should handle write errors 1`] = `[Error: Write error]`;

exports[`writeToFile should remove CDATA tags when present 1`] = `
{
  "message": "<write_to_file_path>cdata.txt</write_to_file_path><status>Success</status>",
  "type": "Reply",
}
`;

exports[`writeToFile should not modify content without CDATA tags 1`] = `
{
  "message": "<write_to_file_path>nocdata.txt</write_to_file_path><status>Success</status>",
  "type": "Reply",
}
`;

exports[`writeToFile should not modify partial CDATA tags 1`] = `
{
  "message": "<write_to_file_path>partial.txt</write_to_file_path><status>Success</status>",
  "type": "Reply",
}
`;
